metricbeat/module/kubernetes/util/kubernetes.go

```diff
@@ -532,113 +796,183 @@ func join(fields ...string) string {
-		namespaceWatcher:  namespaceWatcher,
-		replicasetWatcher: replicasetWatcher,
-		jobWatcher:        jobWatcher,
-	}
-
-	watcher.AddEventHandler(kubernetes.ResourceEventHandlerFuncs{
-		AddFunc: func(obj interface{}) {
-			enricher.Lock()
-			defer enricher.Unlock()
-			update(enricher.metadata, obj.(kubernetes.Resource))
-		},
-		UpdateFunc: func(obj interface{}) {
-			enricher.Lock()
-			defer enricher.Unlock()
-			update(enricher.metadata, obj.(kubernetes.Resource))
-		},
-		DeleteFunc: func(obj interface{}) {
-			enricher.Lock()
-			defer enricher.Unlock()
-			delete(enricher.metadata, obj.(kubernetes.Resource))
-		},
-	})
-
-	return &enricher
-}
+	metricsetName string,
+	resourceName string,
+	resourceWatchers *Watchers,
+	config *kubernetesConfig,
+	updateFunc func(kubernetes.Resource) map[string]mapstr.M,
+	deleteFunc func(kubernetes.Resource) []string,
+	indexFunc func(e mapstr.M) string,
+	log *logp.Logger) *enricher {
+
+	enricher := &enricher{
+		metadata:      map[string]mapstr.M{},
+		index:         indexFunc,
+		updateFunc:    updateFunc,
+		deleteFunc:    deleteFunc,
+		resourceName:  resourceName,
+		metricsetName: metricsetName,
+		config:        config,
+		log:           log,
+	}
 
-func (m *enricher) Start() {
-	m.watchersStartedLock.Lock()
-	defer m.watchersStartedLock.Unlock()
-	if !m.watchersStarted {
-		if m.nodeWatcher != nil {
-			if err := m.nodeWatcher.Start(); err != nil {
-				logp.Warn("Error starting node watcher: %s", err)
-			}
-		}
+	resourceWatchers.lock.Lock()
+	defer resourceWatchers.lock.Unlock()
 
-		if m.namespaceWatcher != nil {
-			if err := m.namespaceWatcher.Start(); err != nil {
-				logp.Warn("Error starting namespace watcher: %s", err)
-			}
-		}
+	watcher := resourceWatchers.watchersMap[resourceName]
+	if watcher != nil {
+		watcher.enrichers[metricsetName] = enricher
 
-		if m.replicasetWatcher != nil {
-			if err := m.replicasetWatcher.Start(); err != nil {
-				logp.Warn("Error starting replicaset watcher: %s", err)
+		// Check if there are past events for this resource and update metadata if there are
+		for key, _ := range watcher.metadataObjects {
+			obj, exists, err := watcher.watcher.Store().GetByKey(key)
+			if err != nil {
+				log.Errorf("Error trying to get the object from the store: %s", err)
+			} else {
+				if exists {
+					newMetadataEvents := enricher.updateFunc(obj.(kubernetes.Resource))
+					// add the new metadata to the watcher received metadata
+					for id, metadata := range newMetadataEvents {
+						enricher.metadata[id] = metadata
+					}
+				}
 			}
 		}
 
-		if m.jobWatcher != nil {
-			if err := m.jobWatcher.Start(); err != nil {
-				logp.Warn("Error starting job watcher: %s", err)
-			}
-		}
+		watcher.watcher.AddEventHandler(kubernetes.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				resourceWatchers.lock.Lock()
+				defer resourceWatchers.lock.Unlock()
+
+				// Add object to the list of metadata objects of this watcher
```

Comment:
```suggestion
				// Add object to the list of metadata objects of this watcher
				// so it can be used by enrichers created after the event is
				// triggered
```